### PR TITLE
MacOS workflow: Install harfbuzz normally again

### PR DESCRIPTION
**Pull request by tobbi**
See
https://github.com/orgs/Homebrew/discussions/4735#discussioncomment-6902146

### CODE REVIEW

**Review by ZhongRuoyu — COMMENTED** [review]
You may need to add

```
brew update
brew reinstall harfbuzz
```

to make sure you're using the newest bottle, instead of the one cached in the GitHub Actions runner image: https://github.com/SuperTux/supertux/actions/runs/6071036504/job/16468259452#step:3:24

The `brew reinstall` step shouldn't be needed when the runner image is updated, or when `harfbuzz` is bumped to a newer version. However, I do recommend that `brew update` be kept, because that makes sure the formula (package) information is up-to-date. (The GitHub runner image [explicitly disables auto-update](https://github.com/actions/runner-images/blob/1a30e20713cfc03483b15cc0105259ed6ef9a8df/images/macos/provision/configuration/environment/bashrc#L27).) At the time of writing, [`cmake` is already at 3.27.4](https://formulae.brew.sh/formula/cmake), while the CI logs still incorrectly suggests that the already-installed [`cmake` 3.27.3 is up to date](https://formulae.brew.sh/formula/cmake).

**Review by ZhongRuoyu — COMMENTED** [review]
Looks like you've hit another GitHub Actions runner image issue. Before that gets resolved, `brew unlink python && brew link --overwrite python` should work around it. (At Homebrew/brew we apply the same workaround: https://github.com/Homebrew/brew/commit/298003b18191eb55f58f19ca1b5d0af45e37632a.)

**Comment by tobbi** [comment]
@ZhongRuoyu  Still fails with the olink error. Any idea?

**Comment by ZhongRuoyu** [comment]
I downloaded the last successful artifact from https://github.com/SuperTux/supertux/actions/runs/6068886641 and had a look at `SuperTux.app/Contents/Resources/Frameworks`, where CPack's `fixup_bundle` copies the dylibs to. It turned out that it's `libSDL2_image` that links with `libjpeg`:

```
libSDL2_image-2.0.0.dylib:
	@executable_path/../Frameworks/libSDL2_image-2.0.0.dylib (compatibility version 601.0.0, current version 601.3.0)
	@executable_path/../Frameworks/libpng16.16.dylib (compatibility version 56.0.0, current version 56.0.0)
	@executable_path/../Frameworks/libjxl.0.8.dylib (compatibility version 0.8.0, current version 0.8.1)
	@executable_path/../Frameworks/libjpeg.8.dylib (compatibility version 8.0.0, current version 8.2.2)
	@executable_path/../Frameworks/libtiff.6.dylib (compatibility version 7.0.0, current version 7.0.0)
	@executable_path/../Frameworks/libavif.15.dylib (compatibility version 15.0.0, current version 15.0.1)
	@executable_path/../Frameworks/libwebp.7.dylib (compatibility version 9.0.0, current version 9.6.0)
	@executable_path/../Frameworks/libsharpyuv.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@executable_path/../Frameworks/libSDL2-2.0.0.dylib (compatibility version 2601.0.0, current version 2601.5.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
```

`sdl2_image` was rebuilt two weeks ago (https://github.com/Homebrew/homebrew-core/commit/7878f28f672d3c748b3ecb3e84a4c4af5cb88daf) because of a dependency update. The rebuilt dylib has relocatable install names (the ones with `@loader_path`) in it, which CPack, again, failed to handle:

```console
$ otool -L `brew --prefix sdl2_image`/lib/libSDL2_image.dylib
/usr/local/opt/sdl2_image/lib/libSDL2_image.dylib:
	/usr/local/opt/sdl2_image/lib/libSDL2_image-2.0.0.dylib (compatibility version 601.0.0, current version 601.3.0)
	@loader_path/../../../../opt/libpng/lib/libpng16.16.dylib (compatibility version 57.0.0, current version 57.0.0)
	@loader_path/../../../../opt/jpeg-xl/lib/libjxl.0.8.dylib (compatibility version 0.8.0, current version 0.8.2)
	@loader_path/../../../../opt/jpeg-turbo/lib/libjpeg.8.dylib (compatibility version 8.0.0, current version 8.3.2)
	@loader_path/../../../../opt/libtiff/lib/libtiff.6.dylib (compatibility version 7.0.0, current version 7.1.0)
	@loader_path/../../../../opt/libavif/lib/libavif.16.dylib (compatibility version 16.0.0, current version 16.0.0)
	@loader_path/../../../../opt/webp/lib/libwebp.7.dylib (compatibility version 9.0.0, current version 9.7.0)
	@loader_path/../../../../opt/sdl2/lib/libSDL2-2.0.0.dylib (compatibility version 2801.0.0, current version 2801.2.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
```

We've temporarily disabled relocatable install names until we figure out a way to resolve similar issues (mostly with CPack). For now, I have dispatched a rebuild job for `sdl2_image`. I'll let you know when it's updated.

**Comment by ZhongRuoyu** [comment]
@tobbi I have rebuilt the affected libraries (that I recognised from the `.pkg` artifact), please try again; no modifications needed.

**Comment by ZhongRuoyu** [comment]
Somehow the updated bottle for `jpeg-xl` was not delivered in https://github.com/SuperTux/supertux/actions/runs/6071589928/job/16474229811#step:3:190 (see checksum update for Big Sur at Homebrew/homebrew-core#131958 -- the old one was downloaded). This could be due to GitHub Pages' caching (we host our formula API there). Would you please try rerunning the jobs @tobbi?

Sorry for the back and forth!

**Comment by ZhongRuoyu** [comment]
`webp` needs to be updated, too.

Actually, let me perform more tests in a separate supertux fork. I'll let you know when it builds.

**Comment by ZhongRuoyu** [comment]
@tobbi That turned out to be an unrelated CPack issue that's been there for a while: https://gitlab.kitware.com/cmake/cmake/-/issues/22557

What happened was: `libwebp` made their `libsharpyuv` requirement private in https://chromium.googlesource.com/webm/libwebp.git/+/31c28db53c6fa3be7026212fdd1526280e3f0f52, first released in 1.3.1. The current GitHub Actions macOS 11 image is still on 1.3.0, but it was updated to the new version by doing `brew update` and `brew install`.

The direct consequence of that was the change in required install names:

(Before)

```console
$ otool -L /usr/local/opt/sdl2_image/lib/libSDL2_image-2.0.0.dylib
/usr/local/opt/sdl2_image/lib/libSDL2_image-2.0.0.dylib:
	/usr/local/opt/sdl2_image/lib/libSDL2_image-2.0.0.dylib (compatibility version 601.0.0, current version 601.3.0)
	/usr/local/opt/libpng/lib/libpng16.16.dylib (compatibility version 56.0.0, current version 56.0.0)
	/usr/local/opt/jpeg-xl/lib/libjxl.0.8.dylib (compatibility version 0.8.0, current version 0.8.1)
	/usr/local/opt/jpeg-turbo/lib/libjpeg.8.dylib (compatibility version 8.0.0, current version 8.2.2)
	/usr/local/opt/libtiff/lib/libtiff.6.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/local/opt/libavif/lib/libavif.15.dylib (compatibility version 15.0.0, current version 15.0.1)
	/usr/local/opt/webp/lib/libwebp.7.dylib (compatibility version 9.0.0, current version 9.6.0)
	/usr/local/opt/webp/lib/libsharpyuv.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/local/opt/sdl2/lib/libSDL2-2.0.0.dylib (compatibility version 2601.0.0, current version 2601.5.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
```

(After -- note that `libsharpyuv` is no longer in the list.)

```console
$ otool -L /usr/local/opt/sdl2_image/lib/libSDL2_image-2.0.0.dylib
/usr/local/opt/sdl2_image/lib/libSDL2_image-2.0.0.dylib:
	/usr/local/opt/sdl2_image/lib/libSDL2_image-2.0.0.dylib (compatibility version 601.0.0, current version 601.3.0)
	/usr/local/opt/libpng/lib/libpng16.16.dylib (compatibility version 57.0.0, current version 57.0.0)
	/usr/local/opt/jpeg-xl/lib/libjxl.0.8.dylib (compatibility version 0.8.0, current version 0.8.2)
	/usr/local/opt/jpeg-turbo/lib/libjpeg.8.dylib (compatibility version 8.0.0, current version 8.3.2)
	/usr/local/opt/libtiff/lib/libtiff.6.dylib (compatibility version 7.0.0, current version 7.1.0)
	/usr/local/opt/libavif/lib/libavif.16.dylib (compatibility version 16.0.0, current version 16.0.1)
	/usr/local/opt/webp/lib/libwebp.7.dylib (compatibility version 9.0.0, current version 9.7.0)
	/usr/local/opt/sdl2/lib/libSDL2-2.0.0.dylib (compatibility version 2801.0.0, current version 2801.2.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
```

CPack's `fixup_bundle` does an iterative search over the required dylibs, so without the direct requirement from `libSDL2_image`, it turned to rely on `libwebp` to find `libsharpyuv`. `libwebp`'s requirements are (in both 1.3.0 and 1.3.1 versions):

```console
otool -L /usr/local/opt/webp/lib/libwebp.7.dylib
/usr/local/opt/webp/lib/libwebp.7.dylib:
	/usr/local/opt/webp/lib/libwebp.7.dylib (compatibility version 9.0.0, current version 9.7.0)
	@rpath/libsharpyuv.0.dylib (compatibility version 1.0.0, current version 1.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
```

And CPack failed to handle the `@rpath`-based install name, as discussed in the CMake issue.

Before this gets fixed on CMake's side, a workaround would be to put this after `webp` gets installed:

```
install_name_tool -change \
  '@rpath/libsharpyuv.0.dylib' \
  /usr/local/opt/webp/lib/libsharpyuv.0.dylib \
  /usr/local/opt/webp/lib/libwebp.7.dylib
```

This effectively changes the `@rpath`-based install name to a full path; CPack does handle that.

**Comment by ZhongRuoyu** [comment]
> Why did this not surface before [actions/runner-images#8078](https://github.com/actions/runner-images/issues/8078) emerged?

I made an incorrect claim in my earlier comment about `webp` version; corrected that. The reason is that `libSDL2_image` was only rebulit recently in https://github.com/Homebrew/homebrew-core/commit/7878f28f672d3c748b3ecb3e84a4c4af5cb88daf. Before that, the reference to `libsharpyuv` was still present.

**Comment by tobbi** [comment]
I'm very confused as to where this leaves us.

**Comment by ZhongRuoyu** [comment]
All that the `install_name_tool` workaround does is (sort-of) "normalise" the install name to a path, just like the other transitive dependencies (see the `libSDL2_image` output above for example). This is not required in usual cases, and its purpose is only to help CPack find the (now-made transitive) dependency `libsharpyuv` of dependency `libwebp`, which CPack is not able to perform currently. Not much change otherwise, as far as I can tell. (One note is that the install names may need to be adjusted when the soversion of either `libwebp` or `libsharpyuv` is bumped, but that's rather uncommon.)

As for when this workaround can be removed: probably not until that CPack issue is resolved. One takeaway from the previous CPack failures (for which Homebrew temporarily disabled relocatable install names) is that CPack does not always handle these less-seen cases.

File: .github/workflows/macos.yml [hunk 1]
```diff
@@ -50,13 +50,16 @@ jobs:
...
         run: |
-          # See https://github.com/orgs/Homebrew/discussions/4735
-          # and https://github.com/actions/runner-images/issues/8078
-          brew uninstall --ignore-dependencies harfbuzz
-          brew install --build-from-source harfbuzz
-          brew install cmake googletest bash rename libogg libvorbis glew\
-                       openal-soft sdl2 sdl2_image sdl2_ttf freetype harfbuzz  \
+          brew update
+          brew unlink python && brew link --overwrite python
+          brew reinstall harfbuzz
```
Comment (ZhongRuoyu) [inline .github/workflows/macos.yml:55]: FYI -- the `macos-11` runner image was last updated yesterday (https://github.com/actions/runner-images/commit/6cc56bac934d65047ef520dce0e8cdfaa20c98c1). Now it should be fine not to reinstall anything.

Comment (tobbi) [reply .github/workflows/macos.yml:55]: Can I get rid of all those lines?

Comment (ZhongRuoyu) [reply .github/workflows/macos.yml:55]: I think it should be fine, yes.

**Review by Vankata453 — APPROVED** [review]
Take this review with a grain of salt, since I do not understand MacOS compilation. However, seems like the workflow finishes, so that's good.

**Comment by tobbi** [comment]
Who *really* had any idea about Mac OS workflows. Hint: I do not

**Comment by Vankata453** [comment]
To be fair, at least you've used MacOS and have compiled ST on it. I haven't. :D

**Comment by mrkubax10** [comment]
> Take this review with a grain of salt

Same, I used macOS only once...

**Comment by tobbi** [comment]
@ZhongRuoyu Thanks for all your help!

**Pull request merged**